src/Idris/Env.idr

```diff
@@ -0,0 +1,46 @@
+module Idris.Env
+
+--- All environment variables accessed by the compiler are enumerated in this module.
+
+import public Data.List.Elem
+
+import System
+
+||| Environment variable used by Idris2 compiler
+public export
+record EnvDesc where
+  constructor MkEnvDesc
+  name : String
+  help : String
+
+||| All environment variables used by Idris2 compiler
+public export
+envs : List EnvDesc
+envs = [
+         MkEnvDesc "EDITOR"        "Editor used in REPL :e command",
+         MkEnvDesc "IDRIS2_PREFIX" "Idris2 installation prefix",
+         MkEnvDesc "IDRIS2_PATH"   "Places Idris2 looks for import files",
+         MkEnvDesc "IDRIS2_DATA"   "Places Idris2 looks for data files",
+         MkEnvDesc "IDRIS2_LIBS"   "Places Idris2 looks for libraries (for code generation)",
+         MkEnvDesc "IDRIS2_CG"     "Idris2 CG directories",
+         MkEnvDesc "CHEZ"          "chez executable used in Chez codegen",
+         MkEnvDesc "RACKET"        "racket executable used in Racket codegen",
+         MkEnvDesc "RACKET_RACO"   "raco executable used in Racket codegen",
+         MkEnvDesc "GAMBIT_GSI"    "gsi executable used in Gambit codegen",
+         MkEnvDesc "GAMBIT_GSC"    "gsc executable used in Gambit codegen",
+         MkEnvDesc "GAMBIT_GSC_BACKEND" "gsc executable backend argument",
+         MkEnvDesc "IDRIS2_CC"     "C compiler executable used in RefC codegen",
+         MkEnvDesc "CC"            "C compiler executable used in RefC codegen",
+         MkEnvDesc "NODE"          "node executable used in Node codegen",
+         MkEnvDesc "PATH"          "PATH variable is used to search for executables in certain codegens"
+       ]
+
+--- `public export` only for `auto` to work in `idrisGetEnv`
+public export
+envNames : List String
+envNames = map (.name) envs
+
+||| Query documented environment variable
+public export
+idrisGetEnv : HasIO io => (name : String) -> {auto known : Elem name Env.envNames} -> io (Maybe String)
```

Comment:
Good idea! I would however use `isYes (isElem name Env.envNames)`
This version will start failing when the list is long enough that
`auto` starts giving up on finding a proof if the element is at
the end. In contrast, `isElem` is a decidability proof and will
not stop until it knows for sure whether the value exists.


We don't seem to have `isYes` defined anywhere yet but it's basically
returning `()` for `Yes` and `Void` for `No`. It'd be worth adding to
the library.

Reply:
I tried this:

```
public export
IsYes : Dec a -> Type
IsYes (Yes prf) = ()
IsYes (No contra) = Void

idrisGetEnv : HasIO io => (name : String) -> {auto known : IsYes (isElem name Env.envNames)} -> io (Maybe String)
```

Error is:

```
Error: While processing right hand side of pathLookup. Can't find an implementation
for IsYes (isElem "PATH" ["EDITOR",
                          "IDRIS2_PREFIX",
                          "IDRIS2_PATH",
                          "IDRIS2_DATA",
                          "IDRIS2_LIBS",
                          "IDRIS2_CG",
                          "CHEZ",
                          "RACKET",
                          "RACKET_RACO",
                          "GAMBIT_GSI",
                          "GAMBIT_GSC",
                          "GAMBIT_GSC_BACKEND",
                          "IDRIS2_CC",
                          "CC",
                          "NODE",
                          "PATH"]).

src/Compiler/Scheme/Chez.idr:35:18--35:36
    |
 35 |     = do path <- idrisGetEnv "PATH"
    |                  ^^^^^^^^^^^^^^^^^^
```

Without `auto` the error is:

```
Error: Unsolved holes:
Compiler.Scheme.Chez.{known:389} introduced at:
src/Compiler/Scheme/Chez.idr:35:18--35:36
    |
 35 |     = do path <- idrisGetEnv "PATH"
    |                  ^^^^^^^^^^^^^^^^^^
```

I'm out ideas what is the issue here. Is that what you meant?

Reply:
Right, you'll need to make [`isElem`](https://github.com/idris-lang/Idris2/blob/master/libs/base/Data/List/Elem.idr#L52) public too.

Reply:
And the [`DecEq` implementation for `String` too](https://github.com/idris-lang/Idris2/blob/master/libs/base/Decidable/Equality.idr#L173)

Reply:
Seems to work. Although I could not return `()`, because it starts complaining like this (and again I have no clue what it means):

```
Error: While processing right hand side of updateEnv. Multiple solutions found in search of:
    ()

src/Idris/Driver.idr:57:30--57:55
    |
 57 |          bdata <- coreLift $ idrisGetEnv "IDRIS2_DATA"
    |                              ^^^^^^^^^^^^^^^^^^^^^^^^^

Possible correct results:
    _ (implicitly bound at src/Idris/Driver.idr:54:10--56:33)
    _ (implicitly bound at src/Idris/Driver.idr:50:10--52:43)
```

My workaround is this:

```
public export
0 IsYes : Dec a -> Type
IsYes (Yes prf) = a
IsYes (No contra) = Void
```

Will submit in a minute.

Reply:
An alternative is to have `IsYes` return a `Bool` and then the
test be `IsYes p === True`.

Reply:
That works. Thanks!